src/cmap/auth/mongodb_oidc/aws_device_workflow.ts

```diff
@@ -0,0 +1,47 @@
+import type { Document } from 'bson';
+import { readFile } from 'fs/promises';
+
+import { MongoAWSError } from '../../../error';
+import type { Connection } from '../../connection';
+import type { MongoCredentials } from '../mongo_credentials';
+import { commandDocument, DeviceWorkflow } from './device_workflow';
+
+/**
+ * Device workflow implementation for AWS.
+ *
+ * @internal
+ */
+export class AwsDeviceWorkflow extends DeviceWorkflow {
+  constructor() {
+    super();
+  }
+
+  /**
+   * Execute the workflow. Looks for AWS_WEB_IDENTITY_TOKEN_FILE in the environment
+   * and then attempts to read the token from that path.
+   */
+  async execute(connection: Connection, credentials: MongoCredentials): Promise<Document> {
+    const token = await getToken();
+    return super.authenticate(connection, credentials, token);
+  }
+
+  /**
+   * Get the document to add for speculative authentication.
+   */
+  async speculativeAuth(): Promise<Document> {
+    const token = await getToken();
+    return { speculativeAuthenticate: commandDocument(token) };
+  }
```

Comment:
I’m (maybe wrongly?) assuming all other device workflows would also follow this pattern – `execute` calls `getToken()` and `super.authenticate()`, `speculativeAuth` calls `getToken()` and wraps it using `commandDocument()`.

Would it make sense to move this logic to `DeviceWorkflow` and instead only have an abstract `getToken()` method?

Reply:
Yeah that's a good suggestion.

Reply:
Updated per suggestion.